From 1f4d952464e525c84ccf4aadd4754862f073a9cb Mon Sep 17 00:00:00 2001
From: mmahesh <mmahesh@maxlinear.com>
Date: Fri, 26 Jul 2024 16:24:55 +0530
Subject: [PATCH] WLANRTSYS-83434 Set correct bandwidth value in user config
 for ACS

ISSUE: 6G vap is coming up with 40MHz in ACS if 320MHz is configured.
ROOTCAUSE: While candidate table population, bandwidth is chosen based
on CONF_OPER enum set, where 320MHz equals to 9.
During ACS, user configured bw (for 320MHz, 4 is given in conf file) is directly taken.
Due to this mismatch, candidate table is populated only till 40MHz.
FIX: User configured bw is converted based on op_class.

FIXUP for WLANRTSYS-50170 Smart ACS implementation
---
 hostapd/ctrl_iface.c     | 2 +-
 hostapd/mxl_ctrl_iface.c | 2 +-
 src/ap/hostapd.c         | 2 +-
 src/ap/mxl_hostapd.c     | 8 +++++---
 src/ap/mxl_hostapd.h     | 2 +-
 5 files changed, 9 insertions(+), 7 deletions(-)

diff --git a/hostapd/ctrl_iface.c b/hostapd/ctrl_iface.c
index 6aedcd50b..ef02b67a1 100644
--- a/hostapd/ctrl_iface.c
+++ b/hostapd/ctrl_iface.c
@@ -1668,7 +1668,7 @@ static int hostapd_ctrl_iface_update_reload(struct hostapd_data *hapd, char *cmd
 #ifdef CONFIG_VENDOR_MXL
 	hapd->iconf->mxl_conf.is_client_mode = true;
 	if (os_strstr(cmd, "bandwidth="))
-		mxl_hostapd_backup_user_conf_cw(iface);
+		mxl_hostapd_update_user_conf_cw(iface);
 #endif /* CONFIG_VENDOR_MXL */
 
 	for (j = 0; j < iface->num_bss; j++) {
diff --git a/hostapd/mxl_ctrl_iface.c b/hostapd/mxl_ctrl_iface.c
index 525562aa8..4d2050fac 100644
--- a/hostapd/mxl_ctrl_iface.c
+++ b/hostapd/mxl_ctrl_iface.c
@@ -3158,7 +3158,7 @@ int mxl_hostapd_ctrl_iface_set(struct hostapd_data *hapd, char *cmd, int *ret)
 		}
 	} else if ((os_strcmp(cmd, "vht_oper_chwidth") == 0) || (os_strcmp(cmd, "he_oper_chwidth") == 0)
 			   || (os_strcmp(cmd, "eht_oper_chwidth") == 0)) {
-		mxl_hostapd_backup_user_conf_cw(hapd->iface);
+		mxl_hostapd_update_user_conf_cw(hapd->iface);
 	/* add new config item here as "} else if...{" */
 	} else
 		return 0;
diff --git a/src/ap/hostapd.c b/src/ap/hostapd.c
index 30189bcea..d2bb4cf63 100644
--- a/src/ap/hostapd.c
+++ b/src/ap/hostapd.c
@@ -4890,7 +4890,7 @@ int hostapd_add_iface(struct hapd_interfaces *interfaces, char *buf)
 	if (start_ctrl_iface(hapd_iface) < 0)
 		goto fail;
 #ifdef CONFIG_VENDOR_MXL
-	mxl_hostapd_backup_user_conf_cw(hapd_iface);
+	mxl_hostapd_update_user_conf_cw(hapd_iface);
 #endif /* CONFIG_VENDOR_MXL */
 	wpa_printf(MSG_INFO, "Add interface '%s'",
 		   hapd_iface->conf->bss[0]->iface);
diff --git a/src/ap/mxl_hostapd.c b/src/ap/mxl_hostapd.c
index 69b9c6338..d5cef1de6 100644
--- a/src/ap/mxl_hostapd.c
+++ b/src/ap/mxl_hostapd.c
@@ -2960,13 +2960,15 @@ void mxl_hostapd_set_conf_from_freq_params(struct hostapd_iface *iface, const st
 	hostapd_set_oper_chwidth(iface->conf, bw);
 }
 
-void mxl_hostapd_backup_user_conf_cw(struct hostapd_iface *iface)
+void mxl_hostapd_update_user_conf_cw(struct hostapd_iface *iface)
 {
-	iface->mxl_iface.acs_user_config_cw = hostapd_get_oper_chwidth(iface->conf);
+	/* Align MXL Proprietary ch_width (4 - 320MHz) according to hostapd (9 - 320MHz) */
+	iface->mxl_iface.acs_user_config_cw = is_6ghz_op_class(iface->conf->op_class) ?
+		op_class_to_ch_width(iface->conf->op_class) : hostapd_get_oper_chwidth(iface->conf);
 	iface->mxl_iface.acs_user_config_op_class = iface->conf->op_class;
 }
 
 void mxl_hostapd_init(struct hostapd_iface * hapd_iface)
 {
-	mxl_hostapd_backup_user_conf_cw(hapd_iface);
+	mxl_hostapd_update_user_conf_cw(hapd_iface);
 }
diff --git a/src/ap/mxl_hostapd.h b/src/ap/mxl_hostapd.h
index 8e3adcc14..97291cb20 100644
--- a/src/ap/mxl_hostapd.h
+++ b/src/ap/mxl_hostapd.h
@@ -211,7 +211,7 @@ int mxl_hostapd_interface_update_power_regd(struct hostapd_data *hapd, int he_6g
 const char * mxl_he_6ghz_pwr_mode_text(enum he_6ghz_ap_type pwr_mode);
 int mxl_hostapd_no_ir_channel_list_updated(struct hostapd_iface *iface);
 int mxl_setup_interface2(struct hostapd_iface *iface);
-void mxl_hostapd_backup_user_conf_cw(struct hostapd_iface *iface);
+void mxl_hostapd_update_user_conf_cw(struct hostapd_iface *iface);
 void mxl_hostapd_init(struct hostapd_iface * hapd_iface);
 
 void mxl_mlo_hostapd_interface_deinit(struct hostapd_iface *iface);
-- 
2.43.0

